components/server/src/orgs/organization-service.ts

```diff
@@ -0,0 +1,82 @@
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { TeamDB } from "@gitpod/gitpod-db/lib";
+import { Organization, TeamMembershipInvite } from "@gitpod/gitpod-protocol";
+import { inject, injectable } from "inversify";
+import { Authorizer } from "../authorization/authorizer";
+import { ErrorCodes, ApplicationError } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { OrganizationPermission } from "../authorization/definitions";
+
+@injectable()
+export class OrganizationService {
+    constructor(
+        @inject(TeamDB) private readonly teamDB: TeamDB,
+        @inject(Authorizer) private readonly auth: Authorizer,
+    ) {}
+    /**
+     * createOrganization creates a new organization
+     */
+    async createOrganization(userId: string, name: string): Promise<Organization> {
+        let result: Organization;
+        try {
+            result = await this.teamDB.transaction(async (db) => {
+                result = await db.createTeam(userId, name);
+                await this.auth.addOrganizationOwnerRole(result.id, userId);
+                return result;
+            });
+        } catch (err) {
+            if (result! && result.id) {
+                await this.auth.removeUserFromOrg(result.id, userId);
+            }
+
+            throw err;
+        }
+        return result;
+    }
+
+    /**
+     * getGenericInvite returns the generic invite for the given organization or creates one if no invite exists.
+     *
+     * @param userId
+     * @param orgID
+     * @returns
+     */
+    public async getOrCreateGenericInvite(userId: string, orgID: string): Promise<TeamMembershipInvite> {
+        const invite = await this.teamDB.findGenericInviteByTeamId(orgID);
+        if (invite) {
+            await this.checkPermissionAndThrow(userId, "invite_members", orgID);
+            if (await this.teamDB.hasActiveSSO(orgID)) {
+                throw new ApplicationError(ErrorCodes.NOT_FOUND, "Invites are disabled for SSO-enabled organizations.");
+            }
+            return invite;
+        }
+        return this.resetGenericInvite(userId, orgID);
+    }
+
+    public async resetGenericInvite(userId: string, orgID: string): Promise<TeamMembershipInvite> {
+        await this.checkPermissionAndThrow(userId, "invite_members", orgID);
+        if (await this.teamDB.hasActiveSSO(orgID)) {
+            throw new ApplicationError(ErrorCodes.NOT_FOUND, "Invites are disabled for SSO-enabled organizations.");
+        }
+        return this.teamDB.resetGenericInvite(orgID);
+    }
+
+    private async checkPermissionAndThrow(userId: string, permission: OrganizationPermission, orgID: string) {
+        if (await this.auth.hasPermissionOnOrganization(userId, permission, orgID)) {
+            return;
+        }
+        // check if the user has read permission
+        if ("read_info" === permission || !(await this.auth.hasPermissionOnOrganization(userId, "read_info", orgID))) {
```

Comment:
Not sure what the problem is with using a specific org permission within the (also specific) org service. I think it makes the code actually more readble to have the `read_info` checkbefore throwing the not found exception.

Reply:
If I understand this correctly, we'd need to then also check `read_members` | `read_sso` etc in these so it'd end up growing to effectively a full list of read permissions against the org entity, no?

Reply:
> Would it make sense to handle these explicitly in each RPC, rather than generally here as that's where the context of the read operation lies, and that's where it is specific with the context?

That would pollute the code with copies of this logic which gives more surface for bugs due to drift

Reply:
I don't think so. Did you find a bug in the logic?

Reply:
> If I understand this correctly, we'd need to then also check read_members | read_sso etc in these so it'd end up growing to effectively a full list of read permissions against the org entity, no?

No, we don't need to enhance this. The `read_info` is needed in order to actually see the resource. If you don't have it you will get NOT FOUND because the resource doesn't exist in your space.

Reply:
```
if (await this.auth.hasPermissionOnOrganization(userId, permission, orgID)) {
            return;
        }
```
This will return false for `write_info`, no? And hence it won't throw when you don't have access.

For `write_sso`, it is kind-of implied that to have `write_sso` you also have `read_info` and hence it would return OK, if you didn't have `write_sso` it returns NotFound, because you ahve `read_info`. I guess I don't like the fact that we're assuming that a given permission (like `write_sso`), is implicitly linked to `read_info`. It kind of breaks the permissions being declarative, as we're reverting to imperative checks

Reply:
Ah, this is that stupid `await` used in an IF.. My bad. It would not trigger the `return` if you don't have perms.

Reply:
This will just return if you have `write_info` and if not check if you have `read_info` if no throws 404 else throws 403. Are we talking about the same code? Do we need to snyc?